Allow the dumping of .dwo files contents to show up when dumping an e…

clayborg opened the pull request:
Allow the dumping of .dwo files contents to show up when dumping an executable with split DWARF.

Currently if you run llvm-dwarfdump on a binary that has skeleton compile units, you only see the skeleton compile units. Since the main binary has the linked addresses it would be nice to be able to dump DWARF from the .dwo files and how the resolved addresses instead of showing the address index and "<unresolved>" in the output. This patch adds an option that can be specified to dump the non skeleton DIEs named --dwo.

Added the ability to use the following options with split dwarf as well:
  --name <name>
  --lookup <addr>
  --debug-info <die-offset>

clayborg committed:
Allow the dumping of .dwo files contents to show up when dumping an executable with split DWARF.

Currently if you run llvm-dwarfdump on a binary that has skeleton compile units, you only see the skeleton compile units. Since the main binary has the linked addresses it would be nice to be able to dump DWARF from the .dwo files and how the resolved addresses instead of showing the address index and "<unresolved>" in the output. This patch adds an option that can be specified to dump the non skeleton DIEs named --dwo.

Added the ability to use the following options with split dwarf as well:
  --name <name>
  --lookup <addr>
  --debug-info <die-offset>

ayermolo commented:
How do you specify to output DIE at an offset for specific non-skeleton unit?

clayborg commented:
> How do you specify to output DIE at an offset for specific non-skeleton unit?

Currently you can just use "--debug-info=<offset>" and you might see multiple DIEs from multiple .dwo files. I can add a ---dwo-id <id> option to allow only dumping a specific .dwo file.

ayermolo commented:
> > How do you specify to output DIE at an offset for specific non-skeleton unit?
> 
> Currently you can just use "--debug-info=" and you might see multiple DIEs from multiple .dwo files. I can add a ---dwo-id option to allow only dumping a specific .dwo file.

right. That is what I was asking. Without it, there will be multiple DIEs being outputted for the same offset.

Not sure a flag entry is necessary.  Can we just re-use --dwo
--dwo --> outputs for all non-skeleton
--dwo=<dwo id> --> for specific ones

adrian-prantl commented:
I think this is useful, the only thing I'm confused about is that the test update seems to only modify whitespace?

dwblaikie commented:
> > > How do you specify to output DIE at an offset for specific non-skeleton unit?
> > 
> > 
> > Currently you can just use "--debug-info=" and you might see multiple DIEs from multiple .dwo files. I can add a ---dwo-id option to allow only dumping a specific .dwo file.
> 
> right. That is what I was asking. Without it, there will be multiple DIEs being outputted for the same offset.
> 
> Not sure a flag entry is necessary. Can we just re-use --dwo 
> --dwo --> outputs for all non-skeleton
> --dwo= --> for specific ones

Sounds good to me. Could allow a comma separated list that could match against the CU hash or the dwo_name? (so you could do `--dwo=0x47,32a32f,foo.dwo` and that could match partial hashes (like git/github will match partial hashes so you don't have to remember too many digits), partial path names, etc))

clayborg commented:
> > > > How do you specify to output DIE at an offset for specific non-skeleton unit?
> > > 
> > > 
> > > Currently you can just use "--debug-info=" and you might see multiple DIEs from multiple .dwo files. I can add a ---dwo-id option to allow only dumping a specific .dwo file.
> > 
> > 
> > right. That is what I was asking. Without it, there will be multiple DIEs being outputted for the same offset.
> > Not sure a flag entry is necessary. Can we just re-use --dwo
> > --dwo --> outputs for all non-skeleton
> > --dwo= --> for specific ones
> 
> Sounds good to me. Could allow a comma separated list that could match against the CU hash or the dwo_name? (so you could do `--dwo=0x47,32a32f,foo.dwo` and that could match partial hashes (like git/github will match partial hashes so you don't have to remember too many digits), partial path names, etc))

I was thinking just adding a `--dwo-id <hash>` option. Then you could do:
```
$ llvm-dwarfdump --dwo-id <hash>  /path/to/dwarf/a.out
```
We could also add a `--dwo-name <match>` option so this could be:
```
$ llvm-dwarfdump --dwo-name foo.dwo  /path/to/dwarf/a.out
```
Any of the above could then be combined with the `--debug-info <die-offset>`

clayborg commented:
And the `--dwo-id` and `--dwo-name` could be partial matches to match parts of the hash or just the basename of the .dwo file

clayborg commented:
The main question is, should we get this patch in as is and then add the new options? Or just combine it all here. 

Also from a testing perspective, how can I add a filename + .dwo test that is portable? The DWO name is usually a full path and I am not sure that relative file paths are supported by the llvm stuff? If we are testing this in llvm, then we can't use clang in the tests right? So I would need to either run obj2yaml on the main executable and .dwo file, but then I still end up with problems with the .dwo path being hard coded

dwblaikie commented:
> The main question is, should we get this patch in as is and then add the new options? Or just combine it all here.

I think figuring out the overall design is a good idea then figure out what the implementation strategy is.

One argument could be made that a "dump everything" mode might not be what we want (if it is what you want, that's OK - we can do that)

I wonder if we didn't do that/didn't add any new flags, and improved the existing search flags 

> Also from a testing perspective, how can I add a filename + .dwo test that is portable?

There's lots of existing testing of Split DWARF (for LLVM output, and for llvm-dwarfdump/llvm-symbolizer/etc) - best to take a look at that for inspiration (from a previous/recent discussion, the sort of preference for implementation strategies these days is to use yaml (yaml2obj to create the inputs), assembly (llvm-mc run in the test to assemble it), or LLVM IR in that order of reduced preference/less isolated testing)

>  The DWO name is usually a full path and I am not sure that relative file paths are supported by the llvm stuff? 

Relative paths are supported and used exclusively at Google, for instance - it's the only way to make portable/directory-independent reproducible builds.

> If we are testing this in llvm, then we can't use clang in the tests right? 

Right

> So I would need to either run obj2yaml on the main executable and .dwo file, but then I still end up with problems with the .dwo path being hard coded

If you were generating this from clang, you can use -fdebug-compilation-dir=. to make the paths relative, or you can modify the assembly or yaml after the fact.

clayborg commented:
> > The main question is, should we get this patch in as is and then add the new options? Or just combine it all here.
> 
> I think figuring out the overall design is a good idea then figure out what the implementation strategy is.
> 
> One argument could be made that a "dump everything" mode might not be what we want (if it is what you want, that's OK - we can do that)
>
> I wonder if we didn't do that/didn't add any new flags, and improved the existing search flags


I usually often do want to dump and entire binary since I often make small example binaries to test with, so I like the idea of being able to dump all of the DWARF since the skeleton DIE on its own isn't very useful, and dumping the .dwo file without the main executable is only useful if you are not not looking at anything that needs addresses. 

I also would like to search be name, by address and by DIE offset, even if that DIE comes from a DWO file. In LLDB the identifier is a uint64_t and many objects in LLDB map back to the DWARF by specifying a uint64_t that contains the DWO index + 32 bit DIE offset. So I often have a function or variable object that has an indentifier that maps back to DWARF using this DWO idx + DIE offset, and I need to find the exact DWARF.

So I would vote that we allow a way to dump the main executable and all of its .dwo files so we can see fully linked addresses. "--dwo" takes are of this currently.

I also have implemented the main search features I want to use with this patch and hooked it up so it all works with .dwo files.

Let me know your thoughts and or suggestions

dwblaikie commented on llvm/lib/DebugInfo/DWARF/DWARFUnit.cpp:
```diff
@@ -751,7 +751,7 @@ void DWARFUnit::updateAddressDieMap(DWARFDie Die) {
   extractDIEsIfNeeded(false);
   if (AddrDieMap.empty())
-    updateAddressDieMap(getUnitDIE());
+    updateAddressDieMap(getNonSkeletonUnitDIE(/*ExtractUnitDIEOnly=*/false));
```
this seems interesting - what motivates this change/could it be separated from the basic "dump the non-dwos" stuff? (`getNonSkeletonUnitDIE` unconditionally could cause a lot more DWARF parsing/trying to read all the DIES in non-skeleton units too when they weren't being read before this patch?)

clayborg replied:
This was motivated by the `--lookup <addr>' stuff. If we have a skeleton compile unit, wouldn't we always want to try and grab the .dwo file and do a real lookup instead of just failing? Or worse yet, we have the .dwo file and we are doing a lookup, but the compiler option that leaves inlines in the skeleton compile unit has been enabled and you get subpar results.

dwblaikie replied:
I'd err on the side of not searching all the dwos without opting in, since it might be quite expensive/a lot of results or things to search through. In any case, seems best to discuss in a separate review, perhaps.

I worry about whether this codepath affects symbolizing in some way, for instance (which the inline info in the skeleton is meant to help with/avoid having to parse the dwos). Separate patches will let us track regressions more carefully.

dwblaikie commented:
> I usually often do want to dump and entire binary since I often make small example binaries to test with, so I like the idea of being able to dump all of the DWARF since the skeleton DIE on its own isn't very useful, and dumping the .dwo file without the main executable is only useful if you are not not looking at anything that needs addresses.

Fair enough - happy enough to add `--dwo` for that use case, then.

> I also would like to search be name, by address and by DIE offset, even if that DIE comes from a DWO file. In LLDB the identifier is a uint64_t and many objects in LLDB map back to the DWARF by specifying a uint64_t that contains the DWO index + 32 bit DIE offset. So I often have a function or variable object that has an indentifier that maps back to DWARF using this DWO idx + DIE offset, and I need to find the exact DWARF.

Yeah, be nice if the existing filters could take, maybe a dwo_name+offset format and an unadorned offset referred to the main file only (since applying it to dwos would be ambiguous, the offset might appear in multiple dwos) and you can use dwo_name+offset, or maybe `*.dwo+offset` for "try to dump this offset in any dwo".

> So I would vote that we allow a way to dump the main executable and all of its .dwo files so we can see fully linked addresses. "--dwo" takes are of this currently.
> 
> I also have implemented the main search features I want to use with this patch and hooked it up so it all works with .dwo files.

Yeah, I'd do those as follow-up patches, with some eye towards a common interface design, possibly the `dwo_name+offset` scheme (eg: `--offset=foo.dwo+0x47`)

clayborg commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -399,6 +408,21 @@ static void filterByName(
+    DWARFDie CUNonSkeletonDie = CU->getNonSkeletonUnitDIE(false);
+    if (CUNonSkeletonDie && CUDie != CUNonSkeletonDie) {
+      // We have a DWO file, we need to search it as well
+      for (const auto &Entry : CUNonSkeletonDie.getDwarfUnit()->dies()) {
+        DWARFDie Die = {CUNonSkeletonDie.getDwarfUnit(), &Entry};
+        if (const char *Name = Die.getShortName())
+          if (filterByName(Names, Die, Name, OS, GetNameForDWARFReg))
+            continue;
+        if (const char *Name = Die.getLinkageName())
+          filterByName(Names, Die, Name, OS, GetNameForDWARFReg);
+      }
```
yep, I can do that.

clayborg committed:
Fixed inline comments and added a test.

clayborg committed:
Remove extra whitespace.

clayborg commented:
Ok, all feedback has been fixed. This patch current only does:
- dumping of the entire .dwo file if the `--dwo` option is specified
- allows lookups by name with `--name <name>` in the .dwo files automatically
- allows lookups by address with `--lookup <addr>` in the .dwo files automatically

Future patches will add the ability to dump specific items by debug info offset in a specific dwo file as @dwblaikie had suggested.

clayborg committed:
run clang format.

dwblaikie commented on llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:
```diff
@@ -390,15 +398,25 @@ static void filterByName(
+    DWARFDie CUDie = CU->getUnitDIE(false);
+    DWARFDie CUNonSkeletonDie = CU->getNonSkeletonUnitDIE(false);
+    // If we have a DWO file, we need to search it as well
+    if (CUNonSkeletonDie && CUDie != CUNonSkeletonDie)
```
Should this be conditional on `DumpNonSkeleton` too? (might be easier to review these changes separately - start with just the `--dwo` flag for plain dumping, then add filtering separately)

clayborg replied:
done

dwblaikie commented on llvm/test/tools/llvm-dwarfdump/dump_dwo.test:
```diff
@@ -0,0 +1,121 @@
+# RUN: yaml2obj %p/Inputs/dump_dwo.o.yaml -o %T/dump_dwo.o
+# RUN: yaml2obj %p/Inputs/dump_dwo.dwo.yaml -o %T/dump_dwo.dwo
+## We need to be in the same directory so we can find .dwo file
+# RUN: cd %T && llvm-dwarfdump --dwo %T/dump_dwo.o | FileCheck %s
+# RUN: cd %T && llvm-dwarfdump --name int --name char %T/dump_dwo.o | FileCheck %s --check-prefix=NAMES
+# RUN: cd %T && llvm-dwarfdump --lookup 0x10 %T/dump_dwo.o | FileCheck %s --check-prefix=LOOKUP
```
Each run line doesn't need its own `cd` - can use one up-front. (maybe even before the `yaml2obj` lines, then those lines don't need to `%T/dump_dwo.o` they can write to `dump_dwo.o`, etc, unqualified.

clayborg replied:
done

clayborg committed:
Fix review comments:

- Revert changes to `DWARFUnit::getSubroutineForAddress(...)`
- Add a "bool CheckDWO" argument to the `DIEsForAddress DWARFContext::getDIEsForAddress(uint64_t Address)` function to allow the DWO file to be searched first if needed since the DWO file usually has more information. The parameter is defaulted to `false`.
- Fix the address lookup calls in llvm-dwarfdump to pass the DumpNonSkeleton bool value to `DIEsForAddress DWARFContext::getDIEsForAddress(uint64_t Address, bool CheckDWO)`
- Add tests that verify --name lookups fail when --dwo is not specified, and succeeds when it is
- Add tests that verify --lookup finds info in the skeleton compile unit when --dwo isn't specified and succeeds when it is in finding info from the split compile unit.

clayborg commented:
Everything should be fixed. Now if you specify `--dwo` along with `--name`, it will search the .dwo files for the name. If `--dwo` isn't specified, then it will search the skeleton compile unit only.

Same thing for lookups. This allows you to search the skeleton compile unit for an address, or the dwo if you specify `--dwo`. 

Added tests for everything.

ayermolo commented:
> > I usually often do want to dump and entire binary since I often make small example binaries to test with, so I like the idea of being able to dump all of the DWARF since the skeleton DIE on its own isn't very useful, and dumping the .dwo file without the main executable is only useful if you are not not looking at anything that needs addresses.
> 
> Fair enough - happy enough to add `--dwo` for that use case, then.
> 
> > I also would like to search be name, by address and by DIE offset, even if that DIE comes from a DWO file. In LLDB the identifier is a uint64_t and many objects in LLDB map back to the DWARF by specifying a uint64_t that contains the DWO index + 32 bit DIE offset. So I often have a function or variable object that has an indentifier that maps back to DWARF using this DWO idx + DIE offset, and I need to find the exact DWARF.
> 
> Yeah, be nice if the existing filters could take, maybe a dwo_name+offset format and an unadorned offset referred to the main file only (since applying it to dwos would be ambiguous, the offset might appear in multiple dwos) and you can use dwo_name+offset, or maybe `*.dwo+offset` for "try to dump this offset in any dwo".
> 
> > So I would vote that we allow a way to dump the main executable and all of its .dwo files so we can see fully linked addresses. "--dwo" takes are of this currently.
> > I also have implemented the main search features I want to use with this patch and hooked it up so it all works with .dwo files.
> 
> Yeah, I'd do those as follow-up patches, with some eye towards a common interface design, possibly the `dwo_name+offset` scheme (eg: `--offset=foo.dwo+0x47`)

This seems over complicates things.
Why not just have --debug-info=0x33 --dwo-id=<dwo id hex> Instead of creating a new option?

clayborg commented:
@dwblaikie Anyting more that needs to happen for this patch? I think I took care of everything that was asked for.

dwblaikie reviewed (approved):
I'd still rather this was submitted without any of the support in the filter/lookup operations, and do those in separate commits each. But it's probably fine.

clayborg commented:
Some buildbots were failing after this due to x86_64 target not being enabled. Fixed with:

commit dc1e2790e476499f6007efcd73ab9d648bfc48bb (HEAD -> main, origin/main, origin/HEAD)
Author: Greg Clayton <clayborg@gmail.com>
Date:   Fri Jan 12 15:02:41 2024 -0800

    Fix buildbots after #66726.
    
    Some builds don't enable the x86_64 target, so we sometimes don't see register names in DWARF output.
